### PR TITLE
[Bug]: `pybamm.Simulation.set_parameters`, `pybamm.Simulation.set_up_and_parameterise_experiment` and `pybamm.Simulation.set_up_and_parameterise_model_for_experiment` made private in `simulation.py`

### LINKED ISSUE #3751: [Bug]: `pybamm.Simulation.set_parameters`, `pybamm.Simulation.set_up_and_parameterise_experiment` and `pybamm.Simulation.set_up_and_parameterise_model_for_experiment` should be private (i.e. begin with underscore _) and not documented
### PyBaMM Version

latest

### Python Version

python3.11

### Describe the bug

Took #3392, This is a part of `API Docs` in this issue.

### Steps to Reproduce

Not applicable

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Akhil-Sharma30**
# Description

`pybamm.Simulation.set_parameters`, `pybamm.Simulation.set_up_and_parameterise_experiment` and `pybamm.Simulation.set_up_and_parameterise_model_for_experiment` made private in `pybamm/simulation.py`

Fixes #3751 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Review by tinosulzer — APPROVED** [review]
Thanks @Akhil-Sharma30

**Review by tinosulzer — CHANGES_REQUESTED** [review]
You will need to also update the method names wherever they are called (which should only be inside the `Simulation` class)

**Review by agriyakhetarpal — APPROVED** [review]
Thanks @Akhil-Sharma30 for the fix!

**Comment by Akhil-Sharma30** [comment]
> You will need to also update the method names wherever they are called (which should only be inside the `Simulation` class)

Done.

**Review by agriyakhetarpal — APPROVED** [review]

**Comment by agriyakhetarpal** [comment]
@all-contributors please add @Akhil-Sharma30 for docs

**Review by arjxn-py — DISMISSED** [review]
Thanks @Akhil-Sharma30, should be good to merge after Valentin's approval.

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Ideally, we should deprecate these functions before making them private, given that they are visible in the API docs, but this might be fine if not a lot of users are using them. However, this should be documented in the CHANGELOG (as a breaking change) and the tests should be fixed (you can look at the logs in GH Actions to see where these functions are being used).

**Comment by Akhil-Sharma30** [comment]
> Ideally, we should deprecate these functions before making them private, given that they are visible in the API docs, but this might be fine if not a lot of users are using them. However, this should be documented in the CHANGELOG (as a breaking change) and the tests should be fixed (you can look at the logs in GH Actions to see where these functions are being used).

Added in the CHANGELOG.md

File: pybamm/simulation.py
Comment (Saransh-cpp) [inline pybamm/simulation.py:191]: Given that not a lot of users are using these methods and we want to make them private in the next release - we could create dummy public methods that throw `NameError` on being invoked -
```python
def set_up_and_parameterise_experiment(self):
    """
    A note here for users that this should not be used and is not private
    with the name _set_up_and_parameterise_experiment
    """
    raise NameError(...)
```

Just suggestions. We should go ahead with the strategy that most maintainers agree with.

Comment (Akhil-Sharma30) [reply pybamm/simulation.py:191]: Should I implement it or wait for the other maintainers feedback on this approach??

Comment (agriyakhetarpal) [reply pybamm/simulation.py:191]: I don't think it's a good idea to make them private right in the next release, since it doesn't conform to our policy of three releases – even if the popularity of this API not being relatively high is the question. Keeping them public plus raising a DeprecationWarning should be enough, a NameError sounds a bit odd IMO (I don't think we have used this before).

File: pybamm/simulation.py
Comment (Saransh-cpp) [inline pybamm/simulation.py:190]:
```suggestion
        msg = "pybamm.simulation.set_up_and_parameterise_experiment is deprecated and not meant to be accessed by users"
```

File: pybamm/simulation.py
Comment (Saransh-cpp) [inline pybamm/simulation.py:216]:
```suggestion
```

File: pybamm/simulation.py
Comment (Saransh-cpp) [inline pybamm/simulation.py:223]:
```suggestion
        msg = "pybamm.simulation.set_up_and_parameterise_model_for_experiment is deprecated not meant to be accessed by users"
```

File: pybamm/simulation.py
Comment (Saransh-cpp) [inline pybamm/simulation.py:335]:
```suggestion
        msg = "pybamm.set_paramters is deprecated and not meant to be accessed by users"
```

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Sorry for taking too long to review, but the deprecation path sounds good to me!

**Comment by Akhil-Sharma30** [comment]
I am not able to reset my branch I made a mess of it. Could someone provide me instruction on how to fix this issue?

**Comment by arjxn-py** [comment]
> I am not able to reset my branch I made a mess of it. Could someone provide me instruction on how to fix this issue?

Can you provide some logs while you're trying to reset your branch?

Although assuming that you might be getting merge conflicts while trying to reset to your branch i.e. :
- I have resolved a merge conflict in `simulation.py` and while doing that i have also removed `import warn` by mistake. I'd now recommend you to make sure that your local working tree is clean before resetting your branch and post logs in case you're still not able to. Plus also add `import warn` again to fix the failing CI.

**Comment by Akhil-Sharma30** [comment]
> > I am not able to reset my branch I made a mess of it. Could someone provide me instruction on how to fix this issue?
> 
> Can you provide some logs while you're trying to reset your branch?
> 
> Although assuming that you might be getting merge conflicts while trying to reset to your branch i.e. :
> 
> * I have resolved a merge conflict in `simulation.py` and while doing that i have also removed `import warn` by mistake. Sorry, I'd now recommend you to make sure that your local working tree is clean before resetting your branch and post logs in case you're still not able to. Plus also add `import warn` again to fix the failing CI.

Thanks for the help. I by mistake deleted the branch but when I was trying to bring the branch locally it was causing issue so I tried the `gh pr checkout 3752` it showed me this error `GraphQL: Could not resolve to a PullRequest with the number of 3752. (repository.pullRequest)`

If possible could you suggest me any method to bring that branch locally again so that I can do the changes.

**Comment by Saransh-cpp** [comment]
I am not very familiar with the GitHub CLI, but with git you should be able to do something like this -
```
git fetch --all   # fetch everything locally (there must be a way to just fetch 1 branch)
git checkout <your_branch>
```

**Comment by brosaplanella** [comment]
Hi @Akhil-Sharma30, any updates on this?

**Comment by kratman** [comment]
@Akhil-Sharma30 This has been open for quite a while are you working on this still? If not we should close it because there have been a lot of changes to PyBaMM since this was opened

**Comment by agriyakhetarpal** [comment]
Could you also resolve the merge conflicts, @Akhil-Sharma30?

**Comment by Akhil-Sharma30** [comment]
> Could you also resolve the merge conflicts, @Akhil-Sharma30?

If possible could you guide me from where can I solve these conflict?

**Comment by kratman** [comment]
> > Could you also resolve the merge conflicts, @Akhil-Sharma30?
> 
> If possible could you guide me from where can I solve these conflict?

@Akhil-Sharma30 Honestly, your change is ~10 lines. It is probably easier to just redo this in a fresh PR. Just check out develop and redo the small changes

**Comment by Akhil-Sharma30** [comment]
> > > Could you also resolve the merge conflicts, @Akhil-Sharma30?
> > 
> > 
> > If possible could you guide me from where can I solve these conflict?
> 
> @Akhil-Sharma30 Honestly, your change is ~10 lines. It is probably easier to just redo this in a fresh PR. Just check out develop and redo the small changes

Okay I have created a fresh PR for this. Just wanted to know if this `pybamm.Simulation.set_up_and_parameterise_experiment` func is removed from the simulation.py?

**Comment by Saransh-cpp** [comment]
Hi, it would be easier to open a new PR and request a review / ask your queries there. We'll be able to help you better in a fresh space. This PR is a bit too tangled now.

**Comment by Akhil-Sharma30** [comment]
I have created a new PR for this fix #4489

**Pull request closed without merging**
